decidim-core/app/helpers/concerns/decidim/user_role_checker.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+module Decidim
+  module UserRoleChecker
+    # Shared behaviour for signed_in admins
+    extend ActiveSupport::Concern
+
+    private
+
+    def user_has_any_role?(user, participatory_space = nil)
+      return false unless user
+      return true if user.admin
+      return true if user.roles.any?
+      return true if participatory_process_user_role?(user, participatory_space)
+      return true if assembly_user_role?(user, participatory_space)
+      return true if conference_user_role?(user, participatory_space)
+
+      false
+    end
+
+    def participatory_process_user_role?(user, participatory_process = nil)
+      return false unless Decidim.module_installed?(:participatory_processes)
+
+      if participatory_process.is_a?(Decidim::ParticipatoryProcess)
+        Decidim::ParticipatoryProcessUserRole.exists?(user:, participatory_process:)
+      else
+        Decidim::ParticipatoryProcessUserRole.exists?(user:)
+      end
+    end
```

Comment:
on the current implementation, it happens that if i log in with a Assembly moderator, i will still be able to post comments in a participatory process. The reason is that the assembly_user_role? will return true as long as i have any roles in any assembly.

This suggestion seems to be fixing the issue.

```suggestion
    def participatory_process_user_role?(user, participatory_process = nil)
      return false unless Decidim.module_installed?(:participatory_processes)
      return false unless participatory_process.is_a?(Decidim::ParticipatoryProcess)

      Decidim::ParticipatoryProcessUserRole.exists?(user:, participatory_process:)
    end
```

Reply:
This is not fixed

Reply:
Thank you for your suggestion. I took some time to implement and test the proposed solution. However, after thorough testing, I found that it didn't work as expected. 

As a result, I've decided to retain the previous implementation for now. I would like to point out that in the `comment_form_spec.rb` file, at line 151, the participatory space of the component being visited is passed. If the user does not have a role in that space, access will be denied.

I'm open to further discussion and any additional insights you might have. Please let me know if you have any other suggestions or if there's anything else you'd like me to try.

Thank you again for your time and feedback!

Reply:
I am not 100% sure what exactly did not went according to plan. The main idea, is that if we allow this check as it is, any Admin (Process, Assembly, Conference) will be able to add comments where they could not, and this has already been determined by Product team that is a bug.